File: backend/lib/summary.js
require('dotenv').config();

const axios = require('axios');

const API_KEY = process.env.BART_API_KEY;
const API_URL = process.env.BART_API_URL;
const MAX_INPUT_LENGTH = 800;
const MAX_SUMMARY_LENGTH = 150;
const MIN_SUMMARY_LENGTH = 50;

const chunkText = (text, chunkSize = MAX_INPUT_LENGTH) => {
    const words = text.split(' ');
    let chunks = [];
    for (let i = 0; i < words.length; i += chunkSize) {
        chunks.push(words.slice(i, i + chunkSize).join(' '));
    }
    return chunks;
};

const summarizeText = async (text, retries = 3) => {
    try {
        const truncatedText = text.length > MAX_INPUT_LENGTH ? text.slice(0, MAX_INPUT_LENGTH) : text;
        const response = await axios.post(
            API_URL,
            {
                inputs: truncatedText,
                parameters: {
                    max_length: MAX_SUMMARY_LENGTH,
                    min_length: MIN_SUMMARY_LENGTH,
                },
            },
            {
                headers: {
                    Authorization: `Bearer ${API_KEY}`,
                },
            }
        );
        if (response.data && response.data[0] && response.data[0].summary_text) {
            return response.data[0].summary_text;
        } else {
            throw new Error('Unexpected response format from summarization API');
        }
    } catch (error) {
        if (error.response && error.response.data.error === 'Model facebook/bart-large-cnn is currently loading' && retries > 0) {
            console.log('Model is loading, retrying...');
            await new Promise(resolve => setTimeout(resolve, 5000));
            return summarizeText(text, retries - 1);
        }
        console.error('Error while summarizing:', error.response?.data || error.message);
        throw new Error('Failed to summarize text');
    }
};

const summarizeTextInChunks = async (text) => {
    const chunks = chunkText(text);
    let summaries = [];
    for (const chunk of chunks) {
        const summary = await summarizeText(chunk);
        summaries.push(summary);
    }
    return summaries.join(' ');
};

module.exports = summarizeTextInChunks;

File: microservice/internal/summary/summary.go
package summary

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	MAX_INPUT_LENGTH   = 800
	MAX_SUMMARY_LENGTH = 150
	MIN_SUMMARY_LENGTH = 50
)

type SummaryResponse struct {
	SummaryText string `json:"summary_text"`
}

func SummarizeText(text string, retries int) (string, error) {
	apiKey := os.Getenv("BART_API_KEY")
	apiURL := os.Getenv("BART_API_URL")
	
	if apiKey == "" || apiURL == "" {
		return "", fmt.Errorf("BART_API_KEY or BART_API_URL not set")
	}

	// Truncate text if too long
	truncatedText := text
	if len(text) > MAX_INPUT_LENGTH {
		truncatedText = text[:MAX_INPUT_LENGTH]
	}

	payload := map[string]interface{}{
		"inputs": truncatedText,
		"parameters": map[string]int{
			"max_length": MAX_SUMMARY_LENGTH,
			"min_length": MIN_SUMMARY_LENGTH,
		},
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal payload: %w", err)
	}

	req, err := http.NewRequest("POST", apiURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read response: %w", err)
	}

	// Check if model is loading
	var errorResp map[string]interface{}
	if err := json.Unmarshal(body, &errorResp); err == nil {
		if errMsg, ok := errorResp["error"].(string); ok {
			if strings.Contains(errMsg, "currently loading") && retries > 0 {
				time.Sleep(5 * time.Second)
				return SummarizeText(text, retries-1)
			}
			return "", fmt.Errorf("API error: %s", errMsg)
		}
	}

	var results []SummaryResponse
	if err := json.Unmarshal(body, &results); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	if len(results) == 0 || results[0].SummaryText == "" {
		return "", fmt.Errorf("unexpected response format from summarization API")
	}

	return results[0].SummaryText, nil
}

func SummarizeTextInChunks(text string) (string, error) {
	chunks := chunkText(text)
	var summaries []string

	for _, chunk := range chunks {
		summary, err := SummarizeText(chunk, 3)
		if err != nil {
			return "", err
		}
		summaries = append(summaries, summary)
	}

	return strings.Join(summaries, " "), nil
}

func chunkText(text string) []string {
	chunkSize := MAX_INPUT_LENGTH
	words := strings.Fields(text)
	var chunks []string

	for i := 0; i < len(words); i += chunkSize {
		end := i + chunkSize
		if end > len(words) {
			end = len(words)
		}
		chunks = append(chunks, strings.Join(words[i:end], " "))
	}

	return chunks
}
